### DISCUSSION TITLE
Problems with relation

### DISCUSSION THREAD

**Discussion by karimpanacci** [prisma/prisma#22180] [Q&A] [answered]
### Question

Hi guys, I'm having problems with this schema:

```prisma
model User {
  id               String             @id @default(cuid())
  name             String
  whatsappAccounts WhatsappAccount[]
  messages         AutomaticMessage[]
  TelegramAccount  TelegramAccount[]
}

model WhatsappAccount {
  phone            String             @id @default(cuid())
  apiKey           String
  User             User               @relation(fields: [userId], references: [id])
  userId           String
  AutomaticMessage AutomaticMessage[]
}

model TelegramAccount {
  phone            String             @id @default(cuid())
  apiKey           String
  User             User               @relation(fields: [userId], references: [id])
  userId           String
  AutomaticMessage AutomaticMessage[]
}

// Messages to send with whatsapp and telegram
model AutomaticMessage {
  id                   String           @id @default(cuid())
  message              String
  whatsappAccount      WhatsappAccount  @relation(fields: [whatsappAccountPhone], references: [phone])
  whatsappAccountPhone String
  TelegramAccount      TelegramAccount  @relation(fields: [telegramAccountPhone], references: [phone])
  telegramAccountPhone String
  userId               String
  User                 User             @relation(fields: [userId], references: [id])
}
```

I would like a constraint that when I create a new AutomaticMessage and select a specific User I am forced to select a whatsappAccount that has exactly the previously user as User.

I can't figure out how to do it? Am I doing something wrong?

**Comment by ludralph** [comment]
Hi @karim2012 👋 

Thank you for raising this question.

In Prisma, relationships are typically defined using the `@relation` attribute, which allows you to specify how models are connected. However, the kind of constraint you're looking for, where creating an `AutomaticMessage` forces a `whatsappAccount` to be selected that is associated with a specific `User`, would require application-level logic or a database trigger rather than something that can be defined directly in the Prisma schema.

You will need to handle this logic in your application code by validating that the `whatsappAccount` belongs to the specified `User` before you create a new `AutomaticMessage`. This could be done by fetching the `whatsappAccount` and checking its `userId` against the `userId` of the `AutomaticMessage` you want to create. For validation or constraints that are not supported by Prisma's schema, you would typically implement this logic in your application's business layer or use a database trigger.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by karimpanacci** [reply]
Hi @ludralph , Thank you for your reply!
So, in your opinion, is there no way to achieve my goal directly from the prisma scheme? Maybe changing the approach with which I try to achieve this goal?

Because doing it at the application level seems very messy to me

**Comment by karimpanacci** [comment]
@ludralph I think i resolved the problem directly in prisma schema, can you take a look at it and give me an opinion?

```prisma
model User {
    id                           String                         @id @default(cuid())
    name                         String
    MessageThirdyPartyConnection MessageThirdyPartyConnection[]
}

model WhatsappAccount {
    phone                        String                         @id @default(cuid())
    apiKey                       String
    MessageThirdyPartyConnection MessageThirdyPartyConnection[]
}

model TelegramAccount {
    phone                        String                         @id @default(cuid())
    apiKey                       String
    MessageThirdyPartyConnection MessageThirdyPartyConnection[]
}

// Messages to send with whatsapp and telegram
model AutomaticMessage {
    id                           String                        @id @default(cuid())
    message                      String
    MessageThirdyPartyConnection MessageThirdyPartyConnection?
}

model MessageThirdyPartyConnection {
    AutomaticMessage   AutomaticMessage @relation(fields: [automaticMessageId], references: [id])
    automaticMessageId String

    TelegramAccount   TelegramAccount @relation(fields: [telegramAccountId], references: [phone])
    telegramAccountId String

    WhatsappAccount   WhatsappAccount @relation(fields: [whatsappAccountId], references: [phone])
    whatsappAccountId String

    User   User   @relation(fields: [userId], references: [id])
    userId String

    @@unique([automaticMessageId])
    @@unique([userId, whatsappAccountId])
    @@unique([userId, telegramAccountId])
    @@unique([userId, whatsappAccountId, telegramAccountId, automaticMessageId])
}
```


UPDATE:
I believe there is a problem in this scheme, because you could create 2 users who have the same whatsapp account...

To solve it you would have to create more relationship tables similar to MessageThirdyPartyConnection... but I think the number of tables would get bigger each time depending on how many `${service}Connection` you add... if for example you also add a smsConnection and instagramConnection table, it becomes unsustainable

Probably the only way is to get your hands on the sql file that prism generates

**Comment by karimpanacci** [comment, marked as answer]
@ludralph Ok... now  i resolved...

```prisma
model User {
    id               String             @id @default(cuid())
    name             String
    WhatsappAccount  WhatsappAccount[]
    TelegramAccount  TelegramAccount[]
    AutomaticMessage AutomaticMessage[]
}

model WhatsappAccount {
    phone            String             @id @default(cuid())
    apiKey           String
    User             User               @relation(fields: [userId], references: [id])
    userId           String
    AutomaticMessage AutomaticMessage[]

    @@unique([phone, userId])
}

model TelegramAccount {
    phone            String             @default(cuid())
    apiKey           String
    User             User               @relation(fields: [userId], references: [id])
    userId           String
    AutomaticMessage AutomaticMessage[]

    @@unique([phone, userId])
}

// Messages to send with whatsapp and telegram
model AutomaticMessage {
    id                   String          @id @default(cuid())
    message              String
    WhatsappAccount      WhatsappAccount @relation(fields: [whatsappAccountPhone, userId], references: [phone, userId])
    whatsappAccountPhone String
    TelegramAccount      TelegramAccount @relation(fields: [telegramAccountPhone, userId], references: [phone, userId])
    telegramAccountPhone String
    User                 User            @relation(fields: [userId], references: [id])
    userId               String
}
```
